contrib/gorm.io/gorm/gorm.go

```diff
@@ -0,0 +1,181 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2020 Datadog, Inc.
+
+// Package gorm provides helper functions for tracing the gorm.io/gorm package (https://github.com/go-gorm/gorm).
+package gorm
+
+import (
+	"context"
+	"database/sql"
+	"math"
+	"time"
+
+	sqltraced "gopkg.in/DataDog/dd-trace-go.v1/contrib/database/sql"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+
+	"gorm.io/gorm"
+)
+
+const (
+	gormConfigKey        = "dd-trace-go:config"
+	gormSpanStartTimeKey = "dd-trace-go:span"
+)
+
+// Open opens a new (traced) database connection. The used driver must be formerly registered
+// using (gopkg.in/DataDog/dd-trace-go.v1/contrib/database/sql).Register.
+func Open(getDialector func(db *sql.DB) gorm.Dialector, driverName, source string, opts ...Option) (*gorm.DB, error) {
```

Comment:
The idea of integrations is that they use the exact same function signatures as the package being integrated with. In this case, a user can just add a new package without changing any of the existing code and have Datadog running with little to no extra code changes.

The current signature to `Open` is the following:
```
func Open(dialector Dialector, config *Config) (db *DB, err error)
```

Reply:
I don't think it's possible to completely mimic the same method signature, since you use sqltrace.Open which needs the driver name, while it's not available in the dialector API. The dialector name can be postgres while the driver name is pgx.

Reply:
What do you suggest to do?

Reply:
What I was suggesting was to move the `sqltraced.Open` part be outside the `Open` function and have the user do that part. Then, the signature could remain the same. Doing that though, I notice that the `Open` function doesn't make much sense anymore so it would likely be better to end up removing it completely and having the user simply call `WithCallbacks` on his own opened database connection (which would probably be better called `Trace` or `RegisterCallbacks`).

I realise this is going now through an experimentation phase to find the best and least intrusive option for a user, but ideally we'd be going through this in the issue before implementing.

What do you think about this?

Reply:
If we ask the user to do the sqltraced.Open himself, then we can also use a plugin like @knusbaum suggested, but then it would require the user to do 2 steps in order to open the DB in a traced manner. I'm not sure which is better.

Reply:
If you don't mind, could you post an example of how that would look like? It doesn't sound too bad based on what you describe, and I think our `jmoiron/sqlx` integration also needs two steps (open, then open again)

Reply:
Sorry for the (very) late response.
I think I can make it work and it will look something like this, let me know what you think:
```
// Register augments the provided driver with tracing, enabling it to be loaded by gormtrace.Open.
sqltrace.Register("pgx", &stdlib.Driver{}, sqltrace.WithServiceName("my-service"))
sqlDb, err := sqltrace.Open("pgx", "postgres://pqgotest:password@localhost/pqgotest?sslmode=disable")
if err != nil {
    log.Fatal(err)
}

db, err := gorm.Open(postgres.New(postgres.Config{Conn: sqlDb}), &gorm.Config{})
if err != nil {
    log.Fatal(err)
}

err = db.Use(gormtrace.NewTracerPlugin())
if err != nil {
    log.Fatal(err)
}
```

Do you think this is better than the current API?

Reply:
@AdallomRoy, Thanks for getting back to this.

To me, this new usage is clearer than a function returning a `gorm.Dialector`.

What do you think about also adding a `gormtrace.Open` with the same signature as `gorm.Open` that calls `gorm.Open` followed by `db.Use(...)`. Then, the example usage would be:
```
sqltrace.Register("pgx", &stdlib.Driver{}, sqltrace.WithServiceName("my-service"))
sqlDb, err := sqltrace.Open("pgx", "postgres://pqgotest:password@localhost/pqgotest?sslmode=disable")
if err != nil {
    log.Fatal(err)
}

db, err := gormtrace.Open(postgres.New(postgres.Config{Conn: sqlDb}), &gorm.Config{})
if err != nil {
    log.Fatal(err)
}
```

Reply:
I updated the API to match your suggestion.
In this case, there is no real need to use gorm Plugins, as this is internal to the tracer anyway, and in my opinion only makes things more complex.
tests & example were fixed to match the new API.

Reply:
I think this looks a lot nicer. I agree, with this method there's no need for a `Plugin`.

The core failure is because `v1.28.0` was released, not something you did. If you upmerge to `v1`, that test should pass.